### DISCUSSION TITLE
Calculating visitation and dispersal metrics on multiple absorption samc

### DISCUSSION THREAD

**Discussion by slamander** [andrewmarx/samc#25] [General]
Hey Andrew,  

Hope all is well with you! 

I had a couple questions about calculating `visitation` and `dispersal` metrics under different absorption states. I have 4 absorption states and I would like to calculate these metrics under each one--independently--using the `*(origin, dest)` parameterization (with roughly 30 origin-destination pairings). When I run visitation, the whole lot takes no longer than 7 minutes, but dispersal takes about 3-4 hours (if it doesn't crash or hang-up indefinitely). 

I noticed that, in your performance vignette, visitation and dispersal under the origin-dest parameterization, computation times should be approximately equivalent. Is that true, or was that an artifact of the old benchmarking tool? 

Currently, I'm creating a new samc object for each absorption scenario and running them separately So, I'm wondering if this is what's causing the long runtimes (e.g., no diagonal caching or excess samc objects in my environment). Should I instead create a multiple absorption samc object and calculate those metrics from? If so, would that give me dispersal/absorption independent of other absorption states? 

If you have any suggestions on how I could make this analysis more efficient, I would be really glad to hear them! 

All my best, 

-Alex.

**Comment by andrewmarx** [comment]
Can you show the code where you run `dispersal()`? As long as there is a `dest` parameter it should be about the same time. If you don't have that parameter, then the longer time makes sense.

**Comment by slamander** [comment]
Sure! I put it in a lapply function to run across the absorption scenarios. I used the exact same code for visitation, just swapping out the function. 
```
disp_mods <- pblapply(abs_scenarios, function(samc){
  dispersal(
    samc = samc,
    origin = origin,
    dest = dest
  ) 
})
```

**Comment by andrewmarx** [comment]
Have you tried creating just a single samc object, running visitation once and timing it, and then running dispersal once and timing it to see if they are similar?

```
# Create samc_obj

start = Sys.time()
r1 = samc::visitation(samc_obj, origin = 1, dest = 100)
print(Sys.time() - start)

start = Sys.time()
r2 = samc::dispersal(samc_obj, origin = 1, dest = 100)
print(Sys.time() - start)
```

My best guess right now is that your timings, especially if it's predicted rather than measured, are being confounded by the caching system. So if you happened to run visitation or certain other metrics before the apply, then when the apply runs, it could be using the cached data and running substantially faster. if you didn't do the same for dispersal and the apply was the first time dispersal is run, then it doesn't have the cached data to use.

I've seen the caching system generally reduce the time for additional runs by about 98%. 7 minutes is about a 97% reduction from 4 hrs, which seems to support the caching hypothesis.
